torch/csrc/jit/codegen/cuda/lower_vectorize_welford.cpp

```diff
@@ -0,0 +1,567 @@
+    // is guaranteed by the fact that the loop is exactly mapped with
+    // a vectorized IterDomain)
+    bool is_predicated = true;
+
+    auto pred = wop_ite->predicate();
+    if (pred->isConst()) {
+      is_predicated = false;
+      // If the conditional is const, it isn't really
+      // predicated. If it's true, this IfThenElse can just be
+      // ignored. If not, the whole block of IfThenElse should be ignored.
+      if (!pred->value()->value()) {
+        // Can this happen? This should be just ignored, assuming
+        // there's no else path.
+        TORCH_INTERNAL_ASSERT(
+            wop_ite->elseBody().empty(),
+            "Unexpected IfThenElse: ",
+            wop_ite->toString());
+      }
+    }
+
+    // The transformation pattern slightly varies depending on how the
+    // Welford expr is predicated. See the comment of each of the
+    // transformation functions.
+
+    if (is_predicated) {
+      // When predicated, check if the whole loop body has the same
+      // predicate. If so, just wrap the loop with the predicate
+      if (canPredicateWholeLoop(wop, wop_ite)) {
+        vectorizeWithLoopPredicate(wop);
+      } else {
+        vectorizeWithInlinePredicate(wop);
+      }
+    } else {
+      vectorizeWithoutPredicate(wop);
+    }
+  }
+
+  // Non-predicated case
+  //
+  // Before:
+  // for () {
+  //   if (true) {
+  //     welfordCombine(...);
+  //   }
+  // }
+  //
+  // After:
+  // nvfuser_index_t new_count = outN()[0] + 1;
+  // float reciprocal = 1 / new_count;
+  // for () {
+  //   welfordVectorized(..., new_count, reciprocal);
+  // }
+  void vectorizeWithoutPredicate(WelfordOp* wop) {
+    auto vectorized_wop = applyVectorizeTransformation(wop, nullptr);
+
+    registerReplace(wop, vectorized_wop);
+  }
+
+  // Predicated case
+  //
+  // Before:
+  // for (i) {
+  //   if (pred) {
+  //     welfordCombine(...);
+  //   }
+  // }
+  //
+  // After:
+  // bool p = pred;
+  // nvfuser_index_t new_count = outN()[0] + p;
+  // float reciprocal = 0;
+  // if (p) {
+  //   reciprocal = 1 / new_count;
+  // }
+  // for (i) {
+  //   welfordVectorized(..., new_count, reciprocal, p);
+  // }
+  void vectorizeWithInlinePredicate(WelfordOp* wop) {
+    kir::IfThenElse* wop_ite = scope_exprs_.back()->as<kir::IfThenElse>();
+    auto conditional = wop_ite->predicate()->value();
+
+    // Allocate a boolean scalar for cond
+    auto pred_var = defineScalar(DataType::Bool)->as<Bool>();
+
+    registerInsertBeforeInnerMostLoop(
+        IrBuilder::create<UnaryOp>(UnaryOpType::Set, pred_var, conditional));
+
+    auto vectorized_wop = applyVectorizeTransformation(wop, pred_var);
+
+    // Replace the predicated welfordOp with welfordVectorized. The
+    // if-then-else predicate is no longer needed.
+    registerReplace(wop_ite, vectorized_wop, *(scope_.rbegin() + 1));
+  }
+
+  // Predicated case when the innermost loop has no externaly
+  // visible effect except for the welford outputs
+  //
+  // Before:
```

Comment:
Is the before supposed to be the same in this case as the inline predicate?

Reply:
Yes, it's checked in `isVectorizableWelford`:

```
   // The predicate should be either Manual or Inline
    if (wop_ite->predicate()->predicate_type() != PredicateType::Manual &&
        wop_ite->predicate()->predicate_type() != PredicateType::Inline) {
      return false;
    }
```